Isthmus To/From SQL examples & enhanced APIs (#625)
substrait-io/substrait-java: examples/isthmus-api/README.md

}
extension_uris {
  extension_uri_anchor: 1
  uri: "/functions_comparison.yaml"
}
extensions {
  extension_function {
    extension_uri_reference: 1
    function_anchor: 1
    name: "equal:any_any"
    extension_urn_reference: 1
  }
}
extensions {
  extension_function {
    extension_uri_reference: 2
    function_anchor: 2
    name: "count:"
    extension_urn_reference: 2
  }
}
relations {....}
}
version {
  minor_number: 77
  producer: "isthmus"
}
extension_urns {
  extension_urn_anchor: 1
  urn: "extension:io.substrait:functions_comparison"
}
extension_urns {
  extension_urn_anchor: 2
  urn: "extension:io.substrait:functions_aggregate_generic"
}

File written to substrait.plan
```

Please see the code comments for details of how the conversion is done.

## Creating SQL from a Substrait Plan

To run [`ToSql.java`](./src/main/java/io/substrait/examples/ToSql.java), execute the command below from the root of this repository. The example reads from `substrait.plan` (likely the file created by `FromSql`) and outputs SQL. The text format of the protobuf has been abbreviated.
```bash
./gradlew examples:isthmus-api:run --args "ToSql substrait.plan"

> 

Reviewer: I suggest splitting the command and sample output into separate code blocks so only the command can be more easily copied.

````suggestion
To run [`ToSql.java`](./src/main/java/io/substrait/examples/ToSql.java), execute the command below from the root of this repository.

```bash
./gradlew examples:isthmus-api:run --args "ToSql substrait.plan"
```

The example reads from `substrait.plan` (likely the file created by `FromSql`) and outputs SQL. The text format of the protobuf has been abbreviated.

```
````

# Isthmus API Examples

The Isthmus library converts Substrait plans to and from SQL Plans. There are two examples showing conversion in each direction.

## How does this work in theory?

> In both cases, the Calcite library is used to do parsing and generation of the SQL String. Calcite has it's own relational object model, so there are classes within Isthmus to convert Substrait to and from Calcites object model.

Reviewer 2: Edited a bit for clarity
```suggestion
The [Calcite](https://calcite.apache.org/) library is used to do parsing and generation of the SQL String. Calcite has it's own relational object model, distinct from substrait's. There are classes within Isthmus to convert Substrait to and from Calcite's object model.
```

Converting to Substrait from SQL uses Calcite to parse the SQL to an object model, and then it will be converted to Substrait.

Converting from Substrait to SQL involves converting Substrait to Calcite's object model, then asking Calcite to generate SQL strings.

## Running the examples

There are 2 example classes:

- [FromSql](./src/main/java/io/substrait/examples/FromSql.java) that creates a plan starting from SQL
- [ToSql](./app/src/main/java/io/substrait/examples/ToSQL.java) that reads a plan and creates the SQL

### Requirements

To run these you will need:

- Java 17 or greater
- [Two datafiles](./app/src/main/resources/) are provided for the sample data

## Creating a Substrait Plan from SQL

> To run [`FromSql.java`](./src/main/java/io/substrait/examples/FromSql.java) from the root of this repository. `subtrait.plan` is the name of file written. It will output to stdout the text format of the protobuf for information; it's quite length so has been abbreviated here.

Reviewer 2: ```suggestion
To run [`FromSql.java`](./src/main/java/io/substrait/examples/FromSql.java), execute the command below from the root of this repository. The example writes a binary plan to `substrait.plan` and outputs the text format of the protobuf to stdout. The output is quite lengthy, so it has been abbreviated here.
```

Again the text format of the protobuf has been abbreviated.
```bash
./gradlew examples:isthmus-api:run --args "ToSql substrait.plan"

> Task :examples:isthmus-api:run
Reading from substrait.plan
extension_uris {
  extension_uri_anchor: 2
  uri: "/functions_aggregate_generic.yaml"
}
extension_uris {
  extension_uri_anchor: 1
  uri: "/functions_comparison.yaml"
}
extensions {
  extension_function {
    extension_uri_reference: 1
    function_anchor: 1
    name: "equal:any_any"
    extension_urn_reference: 1
  }
}
extensions {....}
relations {....}
version {
  minor_number: 77
  producer: "isthmus"
}
extension_urns {
  extension_urn_anchor: 1
  urn: "extension:io.substrait:functions_comparison"
}
extension_urns {
  extension_urn_anchor: 2
  urn: "extension:io.substrait:functions_aggregate_generic"
}

SELECT `t2`.`colour0` AS `COLOUR`, `t2`.`$f1` AS `COLOURCOUNT`
FROM (SELECT `vehicles`.`colour` AS `colour0`, COUNT(*) AS `$f1`
FROM `vehicles`
INNER JOIN `tests` ON `vehicles`.`vehicle_id` = `tests`.`vehicle_id`
WHERE `tests`.`test_result` = 'P'
GROUP BY `vehicles`.`colour`
ORDER BY COUNT(*) IS NULL, 2) AS `t2`

```

> The SQL statement in the selected dialect will be created (MySql is used in the example).

Reviewer 2: How is dialect selected?

Converting to Substrait from SQL uses Calcite to parse the SQL to an object model, and then it will be converted to Substrait.

Converting from Substrait to SQL involves converting Substrait to Calcite's object model, then asking Calcite to generate SQL strings.

## Running the examples

There are 2 example classes:

- [FromSql](./src/main/java/io/substrait/examples/FromSql.java) that creates a plan starting from SQL
- [ToSql](./app/src/main/java/io/substrait/examples/ToSQL.java) that reads a plan and creates the SQL

### Requirements

To run these you will need:

- Java 17 or greater
- [Two datafiles](./app/src/main/resources/) are provided for the sample data

## Creating a Substrait Plan from SQL

To run [`FromSql.java`](./src/main/java/io/substrait/examples/FromSql.java) from the root of this repository. `subtrait.plan` is the name of file written.

```bash
 ./gradlew examples:isthmus-api:run --args "FromSql substrait.plan"
> Task :examples:isthmus-api:run

> Plan{version=Version{major=0, minor=77, patch=0, producer=isthmus}, roots=[Root{input=Sort{input=Aggregate{input=Project{remap=Remap{indices=[15]}, input=Filter{input=Join{left=NamedScan{initialSchema=NamedStruct{struct=Struct{nullable=false, fields=[VarChar{nullable=true, length=15}, VarChar{nullable=true, length=40}, VarChar{nullable=true, length=40}, VarChar{nullable=true, length=15}, VarChar{nullable=true, length=15}, I32{nullable=true}, VarChar{nullable=true, length=15}]}, names=[vehicle_id, make, model, colour, fuel_type, cylinder_capacity, first_use_date]}, names=[vehicles]}, right=NamedScan{initialSchema=NamedStruct{struct=Struct{nullable=false, fields=[VarChar{nullable=true, length=15}, VarChar{nullable=true, length=15}, VarChar{nullable=true, length=20}, VarChar{nullable=true, length=20}, VarChar{nullable=true, length=20}, VarChar{nullable=true, length=15}, I32{nullable=true}, VarChar{nullable=true, length=15}]}, names=[test_id, vehicle_id, test_date, test_class, test_type, test_result, test_mileage, postcode_area]}, names=[tests]}, condition=ScalarFunctionInvocation{declaration=equal:any_any, arguments=[FieldReference{segments=[StructField{offset=0}], type=VarChar{nullable=true, length=15}}, FieldReference{segments=[StructField{offset=8}], type=VarChar{nullable=true, length=15}}], options=[], outputType=Bool{nullable=true}}, joinType=INNER}, condition=ScalarFunctionInvocation{declaration=equal:any_any, arguments=[FieldReference{segments=[StructField{offset=12}], type=VarChar{nullable=true, length=15}}, VarCharLiteral{nullable=false, value=P, length=15}], options=[], outputType=Bool{nullable=true}}}, expressions=[FieldReference{segments=[StructField{offset=3}], type=VarChar{nullable=true, length=15}}]}, groupings=[Grouping{expressions=[FieldReference{segments=[StructField{offset=0}], type=VarChar{nullable=true, length=15}}]}], measures=[Measure{function=AggregateFunctionInvocation{declaration=count:, arguments=[], options=[], aggregationPhase=INITIAL_TO_RESULT, sort=[], outputType=I64{nullable=false}, invocation=ALL}}]}, sortFields=[SortField{expr=FieldReference{segments=[StructField{offset=1}], type=Struct{nullable=false, fields=[VarChar{nullable=true, length=15}, I64{nullable=false}]}}, direction=ASC_NULLS_LAST}]}, names=[COLOUR, COLOURCOUNT]}], expectedTypeUrls=[]}

Author: Yes - I agree; if you look in the other samples there are some utility functions to pretty print plans. 
But that's duplicating code; and @nielspardon pointed out there are other community efforts for text formatting of plans as well. 

Happy to take a follow-on issue to collect the various different formats together to present to the maintainers.

> Plan{version=Version{major=0, minor=77, patch=0, producer=isthmus}, roots=[Root{input=Sort{input=Aggregate{input=Project{remap=Remap{indices=[15]}, input=Filter{input=Join{left=NamedScan{initialSchema=NamedStruct{struct=Struct{nullable=false, fields=[VarChar{nullable=true, length=15}, VarChar{nullable=true, length=40}, VarChar{nullable=true, length=40}, VarChar{nullable=true, length=15}, VarChar{nullable=true, length=15}, I32{nullable=true}, VarChar{nullable=true, length=15}]}, names=[vehicle_id, make, model, colour, fuel_type, cylinder_capacity, first_use_date]}, names=[vehicles]}, right=NamedScan{initialSchema=NamedStruct{struct=Struct{nullable=false, fields=[VarChar{nullable=true, length=15}, VarChar{nullable=true, length=15}, VarChar{nullable=true, length=20}, VarChar{nullable=true, length=20}, VarChar{nullable=true, length=20}, VarChar{nullable=true, length=15}, I32{nullable=true}, VarChar{nullable=true, length=15}]}, names=[test_id, vehicle_id, test_date, test_class, test_type, test_result, test_mileage, postcode_area]}, names=[tests]}, condition=ScalarFunctionInvocation{declaration=equal:any_any, arguments=[FieldReference{segments=[StructField{offset=0}], type=VarChar{nullable=true, length=15}}, FieldReference{segments=[StructField{offset=8}], type=VarChar{nullable=true, length=15}}], options=[], outputType=Bool{nullable=true}}, joinType=INNER}, condition=ScalarFunctionInvocation{declaration=equal:any_any, arguments=[FieldReference{segments=[StructField{offset=12}], type=VarChar{nullable=true, length=15}}, VarCharLiteral{nullable=false, value=P, length=15}], options=[], outputType=Bool{nullable=true}}}, expressions=[FieldReference{segments=[StructField{offset=3}], type=VarChar{nullable=true, length=15}}]}, groupings=[Grouping{expressions=[FieldReference{segments=[StructField{offset=0}], type=VarChar{nullable=true, length=15}}]}], measures=[Measure{function=AggregateFunctionInvocation{declaration=count:, arguments=[], options=[], aggregationPhase=INITIAL_TO_RESULT, sort=[], outputType=I64{nullable=false}, invocation=ALL}}]}, sortFields=[SortField{expr=FieldReference{segments=[StructField{offset=1}], type=Struct{nullable=false, fields=[VarChar{nullable=true, length=15}, I64{nullable=false}]}}, direction=ASC_NULLS_LAST}]}, names=[COLOUR, COLOURCOUNT]}], expectedTypeUrls=[]}

Reviewer 3: @mbwhite had contributed a text format in the substrait-spark example earlier which would have been duplicated initially in this PR. I think the proto classes `toString()` uses the protobuf text format. In protobuf Python libraries there seems to be an explicit pretty print setting. I'm not sure about the protobuf Java library. Maybe you can see whether that exists for the protobuf text format or alternatively use the protobuf JSON format?
